### PR TITLE
Added test to check object size to the api-server

### PR DESCRIPTION
**Pull request by anubhakushwaha**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** 
Helps #47668

**Special notes for your reviewer**:
So incase of `labels` there is a validation of specific label values to be `<63 characters` [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L89)
But no upper limit on the total size of all labels combined.

So I noticed the following errors while changing the sizes :
a) request size > 1MB `etcdserver: request is too large`
b) request size > 2MB `rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max `

I have added tests to check for the same, let me know your thoughts.
@sttts @lavalamp @nikhita 

**Edit 1 :** Incase of both `labels` and `finalizers` there is no limit on overall object.
`finalizers` too have an upper limit of `253` for each value [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L136) but no validation on the whole string array size

### CODE REVIEW

File: test/integration/master/synthetic_master_test.go [hunk 1]
```diff
@@ -225,6 +225,72 @@ func TestStatus(t *testing.T) {
...
+var expected1 = `etcdserver: request is too large`
+var expected2 = `rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max (3489913 vs. 2097152)`
+
+func TestObjectSizeResponses(t *testing.T) {
+
+	_, s, closeFn := framework.RunAMaster(nil)
+	defer closeFn()
+	transport := http.DefaultTransport
+
+	content1, err := ioutil.ReadFile("dummyfile1.txt")
```
Comment (ayushpateria) [inline test/integration/master/synthetic_master_test.go]: I guess instead of having these large files, you could add the labels through a loop.

File: test/integration/master/synthetic_master_test.go [hunk 2]
```diff
@@ -225,6 +225,72 @@ func TestStatus(t *testing.T) {
 	}
 }
 
+var expected1 = `etcdserver: request is too large`
+var expected2 = `rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max (3489913 vs. 2097152)`
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: would at least remove the concrete numbers from the string. They might change.

File: test/integration/master/synthetic_master_test.go [hunk 3]
```diff
@@ -225,6 +225,72 @@ func TestStatus(t *testing.T) {
...
+	_, s, closeFn := framework.RunAMaster(nil)
+	defer closeFn()
+	transport := http.DefaultTransport
+
+	content1, err := ioutil.ReadFile("dummyfile1.txt")
+	if err != nil {
+		t.Fatalf("Unexpected error %v", err)
+	}
+
+	content2, err := ioutil.ReadFile("dummyfile2.txt")
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: what exactly is in these files? (Github does not show them). Are that namespace objects?

File: test/integration/master/synthetic_master_test.go (hunk 1 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @ayushpateria yeah in the first place only I used a loop to get this file, will do that here as well. Thanks for the review, feel free to suggest other ways in which we can create a big object.

File: test/integration/master/synthetic_master_test.go (hunk 2 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @sttts Yeah right, will remove that

File: test/integration/master/synthetic_master_test.go (hunk 3 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @sttts These are the request body that is of huge size, I was thinking of updating this way to generate this body, will update this

File: test/integration/master/synthetic_master_test.go [hunk 4]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+	var temp = `{
+		"apiVersion": "apps/v1beta1",
+		"kind": "Deployment",
+		"metadata": {
+		   "name": "test-deployment2",
+		   "namespace": "default",
+		   "labels": {`
+
+	for i := 0; i < size; i++ {
+		temp += "\"test" + strconv.Itoa(i) + "\" : \"" + val + "\" ,"
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: Better use a `bytes.Buffer`. Constructing strings with `+=` is quadratic in runtime. Compare http://herman.asia/efficient-string-concatenation-in-go.

File: test/integration/master/synthetic_master_test.go [hunk 5]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+			}
+			}
+			}
+			}`
+
+	return temp
+}
+
+func TestObjectSizeResponses(t *testing.T) {
+
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: nit: no newline

File: test/integration/master/synthetic_master_test.go [hunk 6]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+	_, s, closeFn := framework.RunAMaster(nil)
+	defer closeFn()
+	transport := http.DefaultTransport
+
+	var expectedMsgFor1MB = `etcdserver: request is too large`
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: we usually use `:=` instead of `var` for those constants.

File: test/integration/master/synthetic_master_test.go [hunk 7]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+	var expectedMsgFor2MB = `rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max`
+
+	deployment1 := constructBodyByManyLabels("a", 100000)                    // >1 MB file
+	deployment2 := constructBodyByManyLabels("aaaaaaaaaaaaaaaaaaaa", 100000) // >2 MB file
+
+	//deployment3 := constructBodyByManyAnnotations("a", 100000)
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: remove commented debug code.

File: test/integration/master/synthetic_master_test.go [hunk 8]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+		if err != nil {
+			t.Logf("case %v", r)
+			t.Fatalf("unexpected error: %v", err)
+		}
+		func() {
+			resp, err := transport.RoundTrip(req)
+			defer resp.Body.Close()
+			if err != nil {
+				t.Logf("case %v", r)
+				t.Fatalf("unexpected error: %v", err)
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: Just put the Logf output into the Fatalf string.

File: test/integration/master/synthetic_master_test.go [hunk 9]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
...
+		verb            string
+		URL             string
+		body            string
+		expectedMessage string
+	}{
+		{"1 MB", "POST", "/apis/apps/v1beta1/namespaces/" + metav1.NamespaceDefault + "/deployments", deployment1, expectedMsgFor1MB},
+		{"2 MB", "POST", "/apis/apps/v1beta1/namespaces/" + metav1.NamespaceDefault + "/deployments", deployment2, expectedMsgFor2MB},
+	}
+
+	for _, r := range requests {
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: The output is easier to read with subtests, compare: https://blog.golang.org/subtests

File: test/integration/master/synthetic_master_test.go [hunk 10]
```diff
@@ -224,6 +225,101 @@ func TestStatus(t *testing.T) {
 		}
 	}
 }
+func constructBodyByManyLabels(val string, size int) string {
+
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: nit: no newline

**Comment by sttts** [comment]
Looks very good 👍🏼 Left some small comments.

File: test/integration/master/synthetic_master_test.go (hunk 4 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Updated.

File: test/integration/master/synthetic_master_test.go (hunk 6 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Updated

File: test/integration/master/synthetic_master_test.go (hunk 8 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Done

File: test/integration/master/synthetic_master_test.go (hunk 10 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Removed

**Comment by anubhakushwaha** [comment]
@sttts So the behaviour for `labels` and 'finalizers` is same i.e they don't have a validation for checking the total size unlike annotations ([here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L57) )

File: test/integration/master/synthetic_master_test.go (hunk 9 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Ooh thanks for the interesting read, updated.

File: test/integration/master/synthetic_master_test.go [hunk 11]
```diff
@@ -224,6 +225,120 @@ func TestStatus(t *testing.T) {
 		}
 	}
 }
+func constructBody(val string, size int, field string) string {
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: nit: new line above

File: test/integration/master/synthetic_master_test.go [hunk 12]
```diff
@@ -224,6 +225,120 @@ func TestStatus(t *testing.T) {
...
+	buffer.WriteString(`{
+		"apiVersion": "apps/v1beta1",
+		"kind": "Deployment",
+		"metadata": {
+		   "name": "test-deployment2",
+		   "namespace": "default",
+		   `)
+
+	switch field {
+	case "labels":
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: Compare https://stackoverflow.com/questions/14426366/what-is-an-idiomatic-way-of-representing-enums-in-go for enums in Golang (that syntax with `iota`). That's a bit cleaner than free form strings.

Comment (sttts) [reply test/integration/master/synthetic_master_test.go]: Maybe even more elegant: what do you think about creating a Golang Deployment struct object here and then `json.Marshal` it into json? I imagine the code will be nicer, maybe. Maybe worth a try?

File: test/integration/master/synthetic_master_test.go [hunk 13]
```diff
@@ -224,6 +225,120 @@ func TestStatus(t *testing.T) {
...
+func TestObjectSizeResponses(t *testing.T) {
+	_, s, closeFn := framework.RunAMaster(nil)
+	defer closeFn()
+	transport := http.DefaultTransport
+
+	expectedMsgFor1MB := `etcdserver: request is too large`
+	expectedMsgFor2MB := `rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max`
+	expectedMsgForLargeAnnotation := `metadata.annotations: Too long: must have at most 262144 characters`
+
+	deployment1 := constructBody("a", 100000, "labels")                    // >1 MB file
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: those numbers are hard to follow. How about calling them `deploymentMegabyteLabels`, `deploymentTwoMegabytesLabels`, ... ?

File: test/integration/master/synthetic_master_test.go (hunk 12 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @sttts Completely agree with you, will update it.

**Comment by anubhakushwaha** [comment]
@sttts Updated the code, let me know if I can do things better and I would love to work upon them, thanks

File: test/integration/master/synthetic_master_test.go [hunk 14]
```diff
@@ -24,14 +24,17 @@ import (
...
 	"os"
+	"strconv"
 	"strings"
 	"sync"
 	"testing"
 	"time"
 
 	"github.com/ghodss/yaml"
 
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: can we use v1 here?

File: test/integration/master/synthetic_master_test.go [hunk 15]
```diff
@@ -225,6 +228,133 @@ func TestStatus(t *testing.T) {
...
+					Containers: []corev1.Container{
+						{
+							Name:  "foo",
+							Image: "foo",
+						},
+					},
+				},
+			},
+		},
+	}
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: 👍

File: test/integration/master/synthetic_master_test.go [hunk 16]
```diff
@@ -225,6 +228,133 @@ func TestStatus(t *testing.T) {
...
+		{"2 MB", "POST", RequestURL, deployment2, expectedMsgFor2MB},
+		{"1 MB", "POST", RequestURL, deployment3, expectedMsgForLargeAnnotation},
+		{"1 MB", "POST", RequestURL, deployment4, expectedMsgFor1MB},
+		{"2 MB", "POST", RequestURL, deployment5, expectedMsgFor2MB},
+	}
+
+	for _, r := range requests {
+		t.Run(fmt.Sprintf("%s %s %s %s", r.size, r.verb, r.URL, r.body), func(t *testing.T) {
+			bodyBytes := bytes.NewReader([]byte(r.body))
+			req, err := http.NewRequest(r.verb, s.URL+r.URL, bodyBytes)
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: here we could even use the typed client in k8s.io/client-go/kubernetes for apps/v1.Deployment

Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @sttts  Sure thing, will update it.

File: test/integration/master/synthetic_master_test.go (hunk 14 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Yes, will update it, my mistake.

File: test/integration/master/synthetic_master_test.go (hunk 16 above)
Comment (sttts) [reply test/integration/master/synthetic_master_test.go]: Any update on this?

File: test/integration/master/synthetic_master_test.go (hunk 14 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Done

File: test/integration/master/synthetic_master_test.go (hunk 16 above)
Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: @sttts Updated, PTAL.

Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Further from future should I not squash my commits so that just my recent changes are visible? And squash them just at before submitting the PR?

Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Attached [analysis](https://gist.github.com/anubhakushwaha/2095176d8b9733be1658c42fdef0ea68)

Comment (anubhakushwaha) [reply test/integration/master/synthetic_master_test.go]: Attached gist [here](https://gist.github.com/anubhakushwaha/2095176d8b9733be1658c42fdef0ea68)

File: test/integration/master/synthetic_master_test.go [hunk 17]
```diff
@@ -225,6 +229,114 @@ func TestStatus(t *testing.T) {
...
+			annotationsMap[key] = val
+		}
+		deploymentObject.ObjectMeta.Annotations = annotationsMap
+	case "finalizers":
+		finalizerString := []string{}
+		for i := 0; i < size; i++ {
+			finalizerString = append(finalizerString, val)
+		}
+		deploymentObject.ObjectMeta.Finalizers = finalizerString
+	}
```
Comment (sttts) [inline test/integration/master/synthetic_master_test.go]: default: t.Fatalf(...)

**Comment by sttts** [comment]
One nit. Otherwise lgtm.

**Comment by anubhakushwaha** [comment]
@sttts  lgty?

**Comment by sttts** [comment]
@anubhakushwaha 🎉🚀🎈

**Comment by anubhakushwaha** [comment]
Thank you @sttts  @nikhita
